bookdown/05-extending/05-extending-mlr3pipelines.Rmd

```diff
@@ -1,5 +1,470 @@
+```{r 05-extending-018, eval = FALSE}
+    train = function(inputs) {
+      self$state = list()
+      c(inputs, inputs)
+    },
+
+    predict = function(inputs) {
+      c(inputs, inputs)
+    }
+```
+
+#### Putting it Together
+
+The whole definition thus becomes
+
+```{r 05-extending-019}
+PipeOpCopyTwo = R6::R6Class("PipeOpCopyTwo",
+  inherit = PipeOp,
+  public = list(
+    initialize = function(id = "copy.two") {
+      super$initialize(id,
+        input = data.table::data.table(name = "input", train = "*", predict = "*"),
+        output = data.table::data.table(name = c("output1", "output2"),
+                            train = "*", predict = "*")
+      )
+    },
+
+    train = function(inputs) {
+      self$state = list()
+      c(inputs, inputs)
+    },
+
+    predict = function(inputs) {
+      c(inputs, inputs)
+    }
+  )
+)
+```
+
+We can create an instance of our `PipeOp`, put it in a graph, and see what happens when we train it on something:
+
+```{r 05-extending-020}
+poct = PipeOpCopyTwo$new()
+gr = Graph$new()
+gr$add_pipeop(poct)
+
+print(gr)
+
+result = gr$train(task)
+
+str(result)
+```
+
+### Special Case: Preprocessing {#ext-pipe-preproc}
+
+Many PipeOps perform an operation on exactly one `Task`, and return exactly one `Task`. They may even not care about the "Target" / "Outcome" variable of that task, and only do some modification of some input data.
+However, it is usually important to them that the `Task` on which they perform prediction has the same data columns as the `Task` on which they train.
+For these cases, the auxiliary base class `PipeOpTaskPreproc` exists.
+It inherits from `PipeOp` itself, and other PipeOps should use it if they fall in the kind of use-case named above.
+
+When inheriting from `PipeOpTaskPreproc`, one must either implement the `train_task` and `predict_task` functions, or the `train_dt`, `predict_dt` functions, depending on whether wants to operate on a `Task` object or on `data.table`s.
+In the second case, one can optionally also overload the `select_cols` function, which chooses which of the incoming `Task`'s features are given to the `train_dt` / `predict_dt` functions.
+
+The following will show two examples: `PipeOpDropNA`, which removes a `Task`'s rows with missing values during training (and implements `train_task` and `predict_task`), and `PipeOpScale`, which scales a `Task`'s numeric columns (and implements `train_dt`, `predict_dt`, and `select_cols`).
+
+#### Example: `PipeOpDropNA`
+
+Dropping rows with missing values may be important when training a model that can not handle them.
+
+Because `mlr3` `Task`s only contain a view to the underlying data, it is not necessary to modify data to remove rows with missing values. Instead, the rows can be removed using the `Task`'s `$filter` method, which modifies the `Task` in-place. This is done in the `train_task` function. We take care that we also set the `$state` slot to signal that the `PipeOp` was trained.
```

Comment:
```suggestion
Because `mlr3` `Task`s only contain a view to the underlying data, it is not necessary to modify data to remove rows with missing values. 
Instead, the rows can be removed using the `Task`'s `$filter` method, which modifies the `Task` in-place. 
This is done in the `train_task` function. 
We take care that we also set the `$state` slot to signal that the `PipeOp` was trained.
```